RecipeManager.py

```diff
@@ -0,0 +1,21 @@
+from recipe import Recipe
+
+class RecipeManager:
+    def __init__(self):
+        self.recipes = []
+
+    def add_recipe(self, title):
+        self.recipes.append(Recipe(title))
+
+    def remove_recipe(self, index):
```

Comment:
The biggest problem I have is really with the use of index in the context of RecipeManager. At this point I have to go into some detail about an Object Orientated design principle called "Encapsulation":

"Encapsulation can be thought of as putting something inside a capsule — you limit its exposure to the outside world. In software, restricting access to inner objects and properties helps with data integrity."

https://www.freecodecamp.org/news/a-short-overview-of-object-oriented-software-design-c7aa0a622c83/

So what does this mean in this context? Well index is used to reference into the recipes list directly, so whatever calls a function that uses "index" needs to "know" what the correct index is at any given time. Arguably in simple usage it might not be a problem, but whenever you have internal behaviours that can modify parameters whose validity you rely on then it can spell trouble, i.e. the integrity becomes compromised. For example something is removed from a list in multiple places and you make no account for the change and at what point in the list it is relative to. Its one of the reasons why creating "accessors" is a thing, because it can control and orchestrated changes in a way which is internally important to a class, but anything else that uses it should not need to know about or worse try and track.

If the GUI needs to retain a mapping of recipes to places on a display then its the GUI that essentially has to control that because it has to take into account any limitation or changes that the GUI layout or user control makes to the ordering or display of the recipes (e.g. searching or filtering). Its much better in these cases for the GUI to map the features it controls (like the row on a list display) to an object (like a Recipe) in a form which is independent of the internal representation of recipes in the RecipeManager. Typically you would want to make the object representation unique (so a unique field - or UID "unique identifier") in this case at the moment we use the title of the recipe - but this can easily change once we have something basic up and running.

Reply:
I can see how index would make things difficult but as for the GUI I think any function can be slotted in the command= so we can insert the functions from your pull request.
Although I haven't fully wrapped my head around it, I think your paragraph beginning "If the GUI needs to retain a mapping" describes what I've been needing to tie it all together, if we manage to go back to the GUI I think we'd need to implement this.

Reply:
So my presumption was that the GUI would get the list of Recipe classes from the RecipeManager and then display them in a given order, to keep a track of things a list of dictionaries can be created in the GUI and in each dictionary there is the position as the key of the displayed Recipe index position in the displayed list together with the Recipe instance as a value. If you need to refer to a recipe at a particular point in the list you use the index from the display as the key in the dictionary list to find the Recipe instance and then use that to display the detailed recipe for example. If the list of recipes changes for whatever reason the list can be recreated to create the next display (i.e. you iterate the created list to create the next display).